Account for term relationship changes

mboynes opened the pull request:
* Updates to adhere to coding standards
* Adds phpdoc
* Terms are counted when post statuses transition, when term relationships are added, and when term relationships are removed; accounts for the latter two conditions (without duplicating efforts on the former)
* Adds more test coverage
* Simplifies travis builds and tests PHP 7.x

mboynes committed:
Coding standards

mboynes committed:
Only instantiate the class when necessary

mboynes committed:
s/stati/statuses/

mboynes committed:
Account for term relationship changes

mboynes committed:
chmod back to 644

mboynes committed:
Merge remote-tracking branch 'upstream/develop' into develop-refactor

# Conflicts:
#	lightweight-term-count-update.php

mboynes commented:
Looks like it's decrementing by 2 when terms are removed from a post. I'll add a test for this and get that bug resolved.

mboynes committed:
Convert class to singleton to improve testing

mboynes committed:
Add more test cases and clean up test files

mboynes committed:
Fix bugs found in unit tests

mboynes committed:
Add Alley to credits

mboynes committed:
phpcs fixes

mboynes committed:
Speed up travis, cover more versions

mboynes committed:
Fix multisite tests

mboynes committed:
Require WP 4.7+

WP 4.7 adds the `$taxonomy` param to the `added_term_relationship` & `deleted_term_relationships` actions. Working around this isn't worth the effort for old versions of WordPress.

mboynes committed:
Don't run the plugin code in cron or wp-cli tasks

mboynes commented:
@david-binda @mattoperry I had some momentum so I ran with this. Give it a review and if it looks good to you, I have a site I can test it on.

One important point, I bumped the minimum WP version to 4.7 due to a change to the `added_term_relationship` and `deleted_term_relationships` actions that went in in 4.7 (the `$taxonomy` param was added). Working around this is a moderate amount of work to support down to 4.3, and it would be very difficult to support < 4.3 (due to changes in `get_term_by()` introduced in 4.3 that we'd need to account for not having the taxonomy). My opinion is to keep it 4.7+ to keep the code as simple as possible, but if either of you feel strongly otherwise, please let me know.

david-binda commented on lightweight-term-count-update.php:
```diff
@@ -12,79 +12,219 @@
+				'fields' => 'tt_ids',
+			) );
+
+			if ( ! empty( $tt_ids ) ) {
```
[`wp_get_object_terms`](https://developer.wordpress.org/reference/functions/wp_get_object_terms/) may return WP_Error. Thus we should also check for `is_array` here.

mboynes replied:
good catch!

david-binda commented on lightweight-term-count-update.php:
```diff
@@ -12,79 +12,219 @@
+		// For non-attachments, let's check if there are any attachment children
+		// with inherited post status -- if so those will need to be re-counted.
+		if ( 'attachment' !== $post->post_type ) {
+			$attachments = new WP_Query( array(
```
This returns only latest 10 attachments. I'm not sure whether setting the `posts_per_page` should be set to `-1` or whether we should come up with some clever direct SQL. Thoughts?

We should also provide `no_found_rows` set to `true` and `ignore_sticky_posts` set to `true`.

mboynes replied:
> This returns only latest 10 attachments. I'm not sure whether setting the `posts_per_page` should be set to `-1` or whether we should come up with some clever direct SQL. Thoughts?

I was wondering that too. I'm not sure what to do here, to be honest, but I do have some suggestions. To maintain parity with core, we would need to set `'posts_per_page' => -1` to catch all possible attachments, but that could theoretically be worse than `_update_post_term_count()`. Here are a few ideas:

1. We could query just for `'fields' => 'ids'`. All we ever use of the post objects are the `ID` and `post_type`, and we already know the post type is `attachment`. This would keep the query light, but it still means that every attachment could theoretically trigger another query to update the term count, so if a post has 100k attachments we could theoretically be looking at 100,002 total sql queries.
     * A sub-option here would be to refactor the code to be able to pool the attachment queries into one query. That would turn the 100,002 queries down to 1 or 2 queries, where one of them would have 100k+ `tt_id`s in the `IN()` clause.
2. We could do a quick count synchronously for the post and then setup an asynchronous cron task to count attachments, if there are any (or if there are more than, say, 100).
3. Count the attachments, and if the number is high, just let `_update_post_term_count()` run as it usually would. Additionally, we could add a filter to optionally override it, and then people can intervene with custom code if it's a real problem for them, and do something else like (1) or (2) or offload the query to Elasticsearch or something.

(1) or (1a) seem like the best options to me of what I've considered. (2) feels over-engineered for the 0.001%, or whatever hilariously low percentage it would be, of use cases that could be problematic. (3) is a fine option too -- we fixed all the use cases most people will ever encounter, and for that one in a million, we simply left it as-is but opened a door for that site to easily solve their problem.

david-binda commented on lightweight-term-count-update.php:
```diff
@@ -12,79 +12,219 @@
-		$transition_type = $this->transition_type( $new, $old );
+			if ( $attachments->have_posts() ) {
+				foreach ( $attachments->posts as $post ) {
+					$this->transition_post_status( $new_status, $old_status, $post );
```
This is not actually working, since the `post_status` of those attachments is `inherit` and the `transition_type` method called on line 125 is going to return `false` as the `inherit` is not among `counted_statuses` property.

Perhaps we could introduce the `transition_type` param to `transition_post_status` method and only call the `transition_type` method in case it's set to it's default value (eg: `null`)?

mboynes replied:
Perhaps I'm misreading my own code, but I believe this should work. On line 141, we're checking `$new_status` and `$old_status` which come from the `transition_post_status` action for the parent post. As best I can tell, at no point here am I checking the status of the attachment (which you're right, would be `inherit`).

david-binda replied:
I was trying to cover this situation in https://github.com/Automattic/lightweight-term-count-update/pull/2/files#diff-753396fe882316d54c2c4341cc3ab424R231 which is currently failing. Let me dig into why is that.

NB: I have tested all the unit tests on a standard WordPress install w/o the plugin being used and they all are passing.

david-binda commented on lightweight-term-count-update.php:
```diff
@@ -12,79 +12,219 @@
+	 */
+	public function added_term_relationship( $object_id, $tt_id, $taxonomy ) {
+		$post = get_post( $object_id );
+		if ( in_array( $post->post_status, $this->counted_statuses, true ) ) {
```
In case a taxonomy is assigned to a non-post object, eg: Users, Comments, Links (see http://justintadlock.com/archives/2011/10/20/custom-user-taxonomies-in-wordpress) this line produces a Fatal error since we are not checking the return value of `get_post`.

The proper fix, tho, would be implementing the enhancement for non-post objects.

mboynes replied:
According to Justin's blog post, it might be moot to try to implement the enhancement for non-post objects:

> **`update_count_callback`:** You must use a custom function for this because WordPress expects the taxonomy to be for the post object type.

I'll dig into the code to see if this is still relevant, since that blog post is 5+ years old. I'll wait for your failing tests before writing any code regardless.

mboynes replied:
Looks like `_update_generic_term_count()` was added in 3.3 which accounts for non-post objects so I'll plan to implement the enhancement. Truth be told, that function isn't bad since it doesn't have to `JOIN` other tables to account for mixed statuses, but this plugin's query is still lighter.

david-binda replied:
Good find! Thanks for doing the research. I, honestly, haven't dug into it beyond writing the unit test. For v1, we could simply workaround and apply the lighter query just for post_type objects. @mboynes do you see some simple solution or should I dive into it?

mboynes replied:
It's cool, I have it already done :-)

david-binda committed:
Adding more unit-tests covering my feedback in https://github.com/Automattic/lightweight-term-count-update/pull/2/

david-binda reviewed (changes requested):
@mboynes it looks awesome! Thank you for all the hard work.

I have just identified some issues with non-post taxonomies (eg custom taxonomy for users - see http://justintadlock.com/archives/2011/10/20/custom-user-taxonomies-in-wordpress), and in attachments handling. I'm sending a PR with unit-tests covering those cases (see https://github.com/alleyinteractive/lightweight-term-count-update/pull/1 ).

Let me know if you have any questions!

mboynes committed:
Check wp_get_object_terms doesn't return wp_error

mboynes committed:
Merge pull request #1 from Automattic/more-unit-tests-for-develop-refactor

Adding more unit-tests covering my feedback

mboynes committed:
Remove unnecessary unregisters

david-binda commented on lightweight-term-count-update.php:
```diff
@@ -12,79 +12,219 @@
+	 */
+	public function setup() {
+		// Prevent core from counting terms.
+		wp_defer_term_counting( true );
```
I like the idea and it seems to be working fine. I'm just not sure whether we should not check whether the counting is still deferred right before running our SQL queries. While checking whether `wp_defer_term_counting()` is still returning `true` might be a quick solution, it does not seem to be bulletproof.

We could perhaps hook the `edited_term_taxonomy` or `edit_term_taxonomy` actions and check whether they were called for the term and taxonomy we are currently updating before running our updates.

I don't think it's necessary for a v1 of the plugin, since it should be really easy to fix the wrong counts (basically just disable the plugin for a while and everything should get in sync sooner or later).

mboynes replied:
Good ideas, I'm going to move this comment to an issue

mboynes committed:
Handle non-post objects and improve attachments

mboynes commented:
Plugin updated with numerous improvements. We have a failing test that I think I might change -- in short, if an unattached attachment (an attachment with post_parent == 0), the plugin counts it because the attachment is [technically considered public](https://github.com/WordPress/WordPress/blob/master/wp-includes/post.php#L649). _However_, core doesn't correctly calculate this right now. Counting it, therefore, sidesteps a core bug (See [#23530](https://core.trac.wordpress.org/ticket/23530), [#22558](https://core.trac.wordpress.org/ticket/22558)).

mboynes committed:
Fix test to count unattached attachments

mboynes committed:
PHPCS fixes

mboynes committed:
Don't run this plugin when importing

mboynes committed:
Add note about counting unattached attachments

david-binda committed:
Adding edit(ed)_term_taxonomy actions and a unit test for testing that they are being fired for all terms and taxonomies

Since we are updating more tt_ids at once, we need to fire the actions inside a loop. Not the best way to do it, but it seems to be necessary.

The test for testing the action is inspired by https://core.trac.wordpress.org/browser/trunk/tests/phpunit/tests/actions/closures.php when it comes to using the $GLOBALS

mboynes committed:
Merge pull request #2 from Automattic/add-edit-edited-term-actions

Adding edit(ed)_term_taxonomy actions and a unit tests

david-binda commented:
Gah, I'm fixing the PHPCS issues.

david-binda committed:
Fixing PHPCS violations in recent code

mboynes committed:
phpcs fixes

mboynes committed:
Merge pull request #3 from Automattic/add-edit-edited-term-actions

Fixing PHPCS violations in recent code
